presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -164,6 +158,12 @@ private void startLoadSplits()
                         startLoadSplits();
                     }
                 }
+                if (outstandingTasks.decrementAndGet() == 0) {
+                    if (fileIterators.isEmpty() && partitions.isEmpty()) {
+                        hiveSplitSource.finished();
+                        return;
```

Comment:
I guess it needs to be profiled. My theory was that many threads are adding and removing from `fileIterators` so it will be more highly contended, so if `partitions` is non-empty then you can avoid reading from the more highly contended `fileIterators`